commons-mongo/src/main/scala/com/avsystem/commons/mongo/BsonReaderInput.scala

```diff
@@ -67,10 +70,60 @@ class BsonReaderObjectInput private(br: BsonReader) extends ObjectInput {
 
   override def hasNext: Boolean = it.hasNext
   override def nextField(): BsonReaderFieldInput = it.next()
+  private[mongo] def peek(): BsonReaderFieldInput = it.peek()
 }
+
 object BsonReaderObjectInput {
   def startReading(br: BsonReader): BsonReaderObjectInput = {
     br.readStartDocument()
     new BsonReaderObjectInput(br)
   }
 }
+
+/**
+  * Modifies iteration order - moves "_id" field to last position if it was first.
+  * This is a workaround for [[MongoWithIdPolymorphicGenCodec]] where objType needs to be read first.
+  */
+private[commons] class ObjectWithIdBsonInput[ID](objectInput: BsonReaderObjectInput)(implicit idCodec: GenCodec[ID]) extends ObjectInput {
+  private val id: Opt[ID] =
```

Comment:
I don't see why you're decoding it to `ID`. Shouldn't that be `ObjectId`?

Reply:
Yeah, decoding to ObjectId is a lot simpler, but the issue is that Mongo accepts not only ObjectIds as id.

Reply:
OK, but the problem is you're passing `ID` to `BsonValueFieldInput` which isn't supposed to return `ID` in any of its methods.